### ISSUE TITLE
Consider using `pyright` for type checking

### ISSUE THREAD

**Issue by Marigold** [owid/etl#472]
## Proposal

Disclaimer - I'm satisfied with `mypy`, it's working well, I was just curious how does pyright compares in terms of speed.

Running mypy starts taking a long time. It takes more than 90s without mypy cache (in CI or my old laptop). It takes ~10s once cached, however in practice it rarely takes those 10s (perhaps we are using cache wrong or missing some arguments?). Running pyright without cache takes 10s and avoids problems such as same module names (we get around it by type checking steps separately in threads). It would be also consistent with VSCode type checks.

The main disadvantage of pyright is that it requires node and everyone would have to switch. So not worth it in my opinion.

## Todo

- [ ] #599 
- [ ] Switch walden to pyright
- [ ] Switch owid-catalog-py to pyright
- [ ] Switch owid-datautils-py to pyright

**Comment by larsyencken** [comment]
I agree that `mypy` is sluggish, and I personally hate getting different type-checking on the command-line vs in the IDE.

The main pain is that we have devs using three different platforms now (MacOS, Linux, Windows), so if we want to require `pyright` then we should come up with standard instructions for all those platforms to install it.

On MacOS, I'm happy with `brew install pyright`. On Ubuntu, perhaps it's as simple as `sudo apt install npm && sudo npm install -g pyright`. Perhaps Windows WSL is just the same as for Ubuntu? @danyx23 confirm?

**Comment by Marigold** [comment]
Interestingly, we just had a bug recently (that we spent quite some time on) which was caught by pylance, but not by mypy. Mypy also doesn't catch other issues in that module, see

```
mypy etl/steps/data/garden/owid/latest/key_indicators/table_population_density.py
```

e.g. `_sanity_checks` returns None instead of `Table` which is ignored by mypy for some reason.

**Comment by pabloarosado** [comment]
I don't have any experience with `pyright`, but to me it seems that `mypy` is acting very erratically. As I mentioned in [this slack thread](https://owid.slack.com/archives/C0193RW5E2J/p1666602276993599), running `make test` on my computer several times in a row will yield either no errors, or random errors in different places. And then also CI fails on github or buildkite, and I don't know if there really is a problem, or I should simply ignore the errors.

**Comment by Marigold** [comment]
I couldn't find a fix for these random errors and had to make steps typecheck optional (showing errors, but always returning success). That together with the fact that typechecking steps is now noticeably slow raises the importance of migrating to pyright. @larsyencken what do you think?

**Comment by larsyencken** [comment]
Could it be that running mypy in parallel is causing race conditions with its local cache? 

So that we're on the same page, it's totally unacceptable to be working with non-deterministic CI tools. Ignoring that is only ok "whilst we're working on fixing it". So we either need to get `mypy` working by reproducing a broken environment and experimenting together, or we need to switch.

**Comment by larsyencken** [comment]
Mypy being flakey should definitely have its own issue

**Comment by Marigold** [comment]
> So we either need to get mypy working by reproducing a broken environment and experimenting together, or we need to switch.

@larsyencken I spent some time trying to fix it, but I gave up given its non-determinism and long runs. Moreover, even though I fixed some errors, new ones appeared (equally tricky). I also feel like mypy's current runtime is becoming more and more annoying so making a switch sounded like an easier option than debugging mypy running in parallel. Fiona and Pablo were also on board with the switch. What do you think, is this the right time?

**Comment by larsyencken** [comment]
I'm happy to review a PR to introduce `pyright` if you're keen. The main blocker or friction is that it's weird to introduce `node` and a specific NPM package as dependencies for a Python project. If it was only 2x faster than non-parallel non-breaking `mypy`, I would still not do it for simplicity of setup. But Mojmir, you're suggesting it's 10x faster, so let's try it out.

I'd want to see:

- Updates to the README with instructions for installing node on each platform
- A decision on whether or not to version pin `pyright` by install it locally (`node_modules`) or whether to use a global latest version
- `make test` give you a nice error message if `pyright` is not installed
- A matching PR in the `ops` repo that patches the ETL build bootstrap script so that it adds nodejs and pyright to the build server in a nice way
